Pour une lecture améliorée, ce notebook est idéalement lu avec une extension qui gère les niveaux des titres Markdown.  
Par exemple avec l'extension 'Table of Contents' du module 'Nbextensions' de Jupyter.

# Mise en place du notebook

## Import des modules

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import linregress as reg_lin
from mesfonctions import scatter_plot
from mesfonctions import arrondi
from mesfonctions import display_scores

from sklearn import set_config
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import FunctionTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import QuantileTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import Lasso
from sklearn.linear_model import LassoCV
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score
from sklearn.compose import TransformedTargetRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.dummy import DummyRegressor

from graphviz import Source
from sklearn.tree import export_graphviz


In [2]:
# pour développement only - cellule à supprimer du notebook final
import mesfonctions
from importlib import reload  
mesfonctions = reload(mesfonctions)
# from mesfonctions import xxx

## Gestion des paramètres d'affichage de Jupyter

In [140]:
# pour gérer l'affichage des dataframes pandas
pd.options.display.max_rows = 100
pd.options.display.max_columns = 200

# pour gérer l'affichage des arrays numpy
# 230 sur ma config bureau, 160 sur ma config portable
np.set_printoptions(linewidth=160)

# affichage des objets sklearn
set_config(display='diagram')

# pour afficher les gros fichiers sur toute la largeur de l'écran
from IPython.core.display import HTML
HTML("<style>.container { width:95% }</style>")

## Fonctions en cours de développement (à déplacer ultérieurement dans un fichier .py externe)

## Chargement du pickle du df nettoyé

In [83]:
df1 = pd.read_pickle('df1_après_nettoyage_Projet_Seattle.pkl')

# Suppression des lignes pour lesquelles la variable 'SiteEUI(kBtu/sf)' n'est pas renseignée
df1 = df1.dropna(subset=['SiteEUI(kBtu/sf)']).copy()

# Troisième approche du problème : utilisation des surfaces des PropertyUseType, en regroupant les types rares

Exemple de modalité problématique dans la deuxième approche : 'Swimming Pool' est présent dans plusieurs types de propriétés, mais les GFA sont toutes de 0 pour cette modalité de propriété.

## Rechargement du pickle du df nettoyé

In [115]:
df2 = pd.read_pickle('df1_après_nettoyage_Projet_Seattle.pkl')

## Nombre de propriétés renseignées pour chaque type de surface

Liste qui contient les modalités :

In [116]:
liste_modalites_types_1_2_3 = list(df2['LargestPropertyUseType'].dropna().unique()) + list(
    df2['SecondLargestPropertyUseType'].dropna().unique()) + list(
        df2['ThirdLargestPropertyUseType'].dropna().unique())

In [117]:
liste_modalites_types_1_2_3 = sorted(list(set(liste_modalites_types_1_2_3)))   # suppression doublons
liste_modalites_types_1_2_3

['Adult Education',
 'Automobile Dealership',
 'Bank Branch',
 'Bar/Nightclub',
 'College/University',
 'Convenience Store without Gas Station',
 'Convention Center',
 'Courthouse',
 'Data Center',
 'Distribution Center',
 'Enclosed Mall',
 'Fast Food Restaurant',
 'Financial Office',
 'Fire Station',
 'Fitness Center/Health Club/Gym',
 'Food Sales',
 'Food Service',
 'Hospital (General Medical & Surgical)',
 'Hotel',
 'K-12 School',
 'Laboratory',
 'Library',
 'Lifestyle Center',
 'Manufacturing/Industrial Plant',
 'Medical Office',
 'Movie Theater',
 'Multifamily Housing',
 'Museum',
 'Non-Refrigerated Warehouse',
 'Office',
 'Other',
 'Other - Education',
 'Other - Entertainment/Public Assembly',
 'Other - Lodging/Residential',
 'Other - Mall',
 'Other - Public Services',
 'Other - Recreation',
 'Other - Restaurant/Bar',
 'Other - Services',
 'Other - Technology/Science',
 'Other - Utility',
 'Other/Specialty Hospital',
 'Parking',
 'Performing Arts',
 'Personal Services (Health/Bea

Dictionnaire qui contient le nb de propriétés pour un type de surface donnée :

In [118]:
# initialisation
dict_modalites_types_1_2_3 = {x: 0 for x in liste_modalites_types_1_2_3}

In [119]:
dict_modalites_types_1_2_3

{'Adult Education': 0,
 'Automobile Dealership': 0,
 'Bank Branch': 0,
 'Bar/Nightclub': 0,
 'College/University': 0,
 'Convenience Store without Gas Station': 0,
 'Convention Center': 0,
 'Courthouse': 0,
 'Data Center': 0,
 'Distribution Center': 0,
 'Enclosed Mall': 0,
 'Fast Food Restaurant': 0,
 'Financial Office': 0,
 'Fire Station': 0,
 'Fitness Center/Health Club/Gym': 0,
 'Food Sales': 0,
 'Food Service': 0,
 'Hospital (General Medical & Surgical)': 0,
 'Hotel': 0,
 'K-12 School': 0,
 'Laboratory': 0,
 'Library': 0,
 'Lifestyle Center': 0,
 'Manufacturing/Industrial Plant': 0,
 'Medical Office': 0,
 'Movie Theater': 0,
 'Multifamily Housing': 0,
 'Museum': 0,
 'Non-Refrigerated Warehouse': 0,
 'Office': 0,
 'Other': 0,
 'Other - Education': 0,
 'Other - Entertainment/Public Assembly': 0,
 'Other - Lodging/Residential': 0,
 'Other - Mall': 0,
 'Other - Public Services': 0,
 'Other - Recreation': 0,
 'Other - Restaurant/Bar': 0,
 'Other - Services': 0,
 'Other - Technology/Scien

In [120]:
def compte_type_propriete(categorie):
    for indice, modalite, surface in zip(df2[categorie].index,   # supprimer itérateur indice
                                         df2[categorie],
                                         df2[categorie+'GFA']):
        if ~np.isnan(surface):
            if surface != 0:
                dict_modalites_types_1_2_3[modalite] +=1
    
compte_type_propriete('LargestPropertyUseType')
compte_type_propriete('SecondLargestPropertyUseType')
compte_type_propriete('ThirdLargestPropertyUseType')


In [121]:
# Tri du résultat
{k: v for k, v in sorted(dict_modalites_types_1_2_3.items(), key=lambda item: item[1])}

{'Swimming Pool': 0,
 'Convention Center': 1,
 'Enclosed Mall': 1,
 'Fast Food Restaurant': 1,
 'Fire Station': 1,
 'Police Station': 1,
 'Residential Care Facility': 1,
 'Wholesale Club/Supercenter': 1,
 'Courthouse': 2,
 'Food Sales': 2,
 'Lifestyle Center': 2,
 'Other - Technology/Science': 2,
 'Vocational School': 2,
 'Convenience Store without Gas Station': 3,
 'Other - Utility': 3,
 'Prison/Incarceration': 3,
 'Adult Education': 4,
 'Bar/Nightclub': 4,
 'Library': 4,
 'Movie Theater': 4,
 'Other - Mall': 4,
 'Personal Services (Health/Beauty, Dry Cleaning, etc)': 4,
 'Urgent Care/Clinic/Other Outpatient': 4,
 'Museum': 5,
 'Other - Public Services': 5,
 'Other/Specialty Hospital': 5,
 'Performing Arts': 5,
 'Other - Lodging/Residential': 6,
 'Automobile Dealership': 7,
 'Strip Mall': 7,
 'Other - Education': 8,
 'Financial Office': 9,
 'Other - Restaurant/Bar': 9,
 'Food Service': 10,
 'Repair Services (Vehicle, Shoe, Locksmith, etc)': 10,
 'Hospital (General Medical & Surgical)'

## Feature engineering

Pour chaque modalité de PropertyUseType, on ajoute une colonne au dataframe df2.  
Cette colonne contiendra la superficie pour la modalité considérée.  
Cependant, les types de propriété avec moins de N occurences seront regroupés dans la variable 'Rare'.  

In [122]:
N = 12

##### Ajout de colonne à df2 pour chaque PropertyUseType avec >N occurences

In [123]:
# Création d'une liste contenant le nom des colonnes créés
liste_variables_PropertyUseTypes_GFA = []

for moda in liste_modalites_types_1_2_3:
    if dict_modalites_types_1_2_3[moda] > N :
        nom_moda_GFA = moda + ' GFA'
        df2[nom_moda_GFA] = 0   # initialisation à 0 sq/f
        liste_variables_PropertyUseTypes_GFA.append(nom_moda_GFA)
    
# Colonne 'Rare GFA'
liste_variables_PropertyUseTypes_GFA.append('Rare GFA')
df2['Rare GFA'] = 0

In [124]:
df2.sample(100)

OSEBuildingID        BuildingType          PrimaryPropertyType  ZipCode  \
1281          21438      NonResidential  Small- and Mid-Sized Office  98109.0   
1931          23877   SPS-District K-12                  K-12 School  98108.0   
176             283      NonResidential  Small- and Mid-Sized Office  98125.0   
568             762      NonResidential                 Retail Store  98101.0   
385             540      NonResidential                    Warehouse  98108.0   
912           20212      NonResidential           Mixed Use Property  98107.0   
1161          21125      NonResidential  Small- and Mid-Sized Office  98103.0   
2682          26593      NonResidential  Small- and Mid-Sized Office  98104.0   
530             717  Nonresidential COS                    Warehouse  98134.0   
1508          22139      NonResidential                 Retail Store  98118.0   
1664          23040      NonResidential                  K-12 School  98106.0   
1330          21535      NonResidential                        Hotel  98109.0   
1992          24085      NonResidential                        Other  98134.0   
2644          26396      NonResidential             Worship Facility  98105.0   
296             426      NonResidential  Small- and Mid-Sized Office  98121.0   
1292          21465  Nonresidential COS                        Other  98109.0   
274             403      NonResidential                 Large Office  98101.0   
2024          24170      NonResidential  Small- and Mid-Sized Office  98115.0   
785           19858      NonResidential           Mixed Use Property  98104.0   
1760          23397      NonResidential                        Other  98106.0   
2034          24202      NonResidential                    Warehouse  98134.0   
3097          29612      NonResidential                  K-12 School  98105.0   
309             440      NonResidential           Mixed Use Property  98109.0   
256             383      NonResidential                 Large Office  98104.0   
512             696      NonResidential                 Large Office  98119.0   
1019          20522      NonResidential                    Warehouse  98144.0   
544             735      NonResidential                    Warehouse  98199.0   
1518          22223      NonResidential                  K-12 School  98107.0   
171             278      NonResidential        Senior Care Community  98104.0   
461             622      NonResidential                        Other  98134.0   
989           20433      NonResidential                        Other  98116.0   
402             558      NonResidential  Supermarket / Grocery Store  98155.0   
1010          20500      NonResidential                    Warehouse  98108.0   
2789          26956      NonResidential                   University  98121.0   
1130          20987      NonResidential  Small- and Mid-Sized Office  98108.0   
340             477              Campus                        Other  98109.0   
2454          25694      NonResidential                        Other  98105.0   
2775          26906      NonResidential               Residence Hall  98105.0   
2614          26254      NonResidential  Small- and Mid-Sized Office  98122.0   
2330          25275      NonResidential       Refrigerated Warehouse  98134.0   
2045          24244      NonResidential             Worship Facility  98116.0   
2120          24532      NonResidential                    Warehouse  98108.0   
145             233  Nonresidential COS                        Other  98109.0   
411             570      NonResidential  Small- and Mid-Sized Office  98102.0   
473             635      NonResidential  Small- and Mid-Sized Office  98134.0   
3163          49697      NonResidential           Mixed Use Property  98104.0   
1669          23056      NonResidential                        Other  98109.0   
419             578      NonResidential          Distribution Center  98106.0   
125             200      NonResidenti

In [125]:
liste_variables_PropertyUseTypes_GFA

['Bank Branch GFA',
 'College/University GFA',
 'Data Center GFA',
 'Distribution Center GFA',
 'Fitness Center/Health Club/Gym GFA',
 'Hotel GFA',
 'K-12 School GFA',
 'Laboratory GFA',
 'Medical Office GFA',
 'Multifamily Housing GFA',
 'Non-Refrigerated Warehouse GFA',
 'Office GFA',
 'Other GFA',
 'Other - Entertainment/Public Assembly GFA',
 'Other - Recreation GFA',
 'Other - Services GFA',
 'Parking GFA',
 'Refrigerated Warehouse GFA',
 'Residence Hall/Dormitory GFA',
 'Restaurant GFA',
 'Retail Store GFA',
 'Self-Storage Facility GFA',
 'Senior Care Community GFA',
 'Social/Meeting Hall GFA',
 'Supermarket/Grocery Store GFA',
 'Worship Facility GFA',
 'Rare GFA']

##### Ajout des surfaces des PropertyUseType dans le dataframe df2

In [126]:
def ajout_surfaces_2(categorie):
    for indice, modalite, surface in zip(df2[categorie].index,
                                         df2[categorie],
                                         df2[categorie+'GFA']):
        if ~np.isnan(surface):
            if modalite + ' GFA' in liste_variables_PropertyUseTypes_GFA:
                df2.at[indice, modalite + ' GFA'] += surface
            else:
                df2.at[indice, 'Rare GFA'] += surface
    
ajout_surfaces_2('LargestPropertyUseType')
ajout_surfaces_2('SecondLargestPropertyUseType')
ajout_surfaces_2('ThirdLargestPropertyUseType')

In [127]:
df2.sample(100, random_state=0)

OSEBuildingID        BuildingType          PrimaryPropertyType  ZipCode  \
2738          26761      NonResidential  Small- and Mid-Sized Office  98105.0   
1839          23562      NonResidential           Mixed Use Property  98104.0   
1139          21088      NonResidential  Small- and Mid-Sized Office  98103.0   
1043          20613      NonResidential                    Warehouse  98134.0   
2076          24374      NonResidential                 Retail Store  98118.0   
921           20237      NonResidential                 Retail Store  98105.0   
444             605      NonResidential  Small- and Mid-Sized Office  98109.0   
2411          25554      NonResidential                        Other  98104.0   
1734          23277      NonResidential                    Warehouse  98134.0   
568             762      NonResidential                 Retail Store  98101.0   
863           20050      NonResidential                 Retail Store  98105.0   
277             406      NonResidential                 Large Office  98101.0   
1557          22420      NonResidential                        Other  98107.0   
552             745      NonResidential                 Large Office  98109.0   
420             579      NonResidential                    Warehouse  98106.0   
2168          24711      NonResidential  Small- and Mid-Sized Office  98104.0   
235             359  Nonresidential COS                 Large Office  98104.0   
479             642      NonResidential                    Warehouse  98134.0   
2022          24164   SPS-District K-12                  K-12 School  98106.0   
2583          26154      NonResidential  Small- and Mid-Sized Office  98119.0   
1281          21438      NonResidential  Small- and Mid-Sized Office  98109.0   
1783          23416      NonResidential           Mixed Use Property  98104.0   
913           20217   SPS-District K-12                  K-12 School  98115.0   
246             371      NonResidential           Mixed Use Property  98101.0   
1278          21433  Nonresidential COS                        Other  98109.0   
77              114      NonResidential                 Large Office  98109.0   
2106          24488      NonResidential  Small- and Mid-Sized Office  98105.0   
1671          23065      NonResidential               Medical Office  98133.0   
1693          23120      NonResidential                        Other  98108.0   
3182          49725      NonResidential                        Other  98101.0   
108             175   SPS-District K-12                  K-12 School  98122.0   
964           20372      NonResidential                        Hotel  98121.0   
2143          24589      NonResidential                    Warehouse  98134.0   
348             490      NonResidential  Small- and Mid-Sized Office  98119.0   
1692          23117              Campus                        Other  98108.0   
2564          26119      NonResidential                 Retail Store  98122.0   
3371          50222  Nonresidential COS                       Office      NaN   
591             792      NonResidential  Small- and Mid-Sized Office  98134.0   
488             657      NonResidential           Mixed Use Property  98121.0   
683           19572      NonResidential                    Warehouse  98118.0   
1746          23299      NonResidential  Small- and Mid-Sized Office  98104.0   
1165          21138      NonResidential           Mixed Use Property  98103.0   
81              118   SPS-District K-12                  K-12 School  98106.0   
2720          26726      NonResidential           Mixed Use Property  98104.0   
178             285      NonResidential             Worship Facility  98105.0   
1884          23689      NonResidential             Worship Facility  98106.0   
1435          21793      NonResidential  Small- and Mid-Sized Office  98108.0   
431             590      NonResidential                    Warehouse  98134.0   
1893          23716      NonResidenti

## Train test split 

POUR LE NOTEBOOK FINAL IL FAUDRA FAIRE LE SPLIT PLUS TOT (avant le nettoyage)  
IL FAUDRA FAIRE LE MEME NETTOYAGE SUR LES JEUX TRAIN ET TEST

POUR L'INSTANT ON NE TRAVAILLE QU'A LA PREDICTION DE SiteEUI(kBtu/sf)

In [128]:
df2 = df2.dropna(subset=['SiteEnergyUse(kBtu)']).copy()

##### Variable à prédire :

In [129]:
y4 = df2[['SiteEnergyUse(kBtu)']].copy()

##### Variables explicatives :

In [130]:
X4 = df2[liste_variables_PropertyUseTypes_GFA].copy()

##### Split :

In [131]:
X_train_EU, X_test_EU, y_train_EU, y_test_EU = train_test_split(X4, y4, test_size=0.2, random_state=0)

## Régression linéaire simple, sans rescaling, cross-validation

In [132]:
lin_reg = LinearRegression();

In [133]:
lin_scores = cross_val_score(lin_reg, X_train_EU, y_train_EU,
                             scoring="neg_root_mean_squared_error", cv=10)
display_scores(-lin_scores, puissance10=6)

Scores: 7.02e6 8.80e6 11.90e6 6.20e6 10.04e6 6.30e6 16.50e6 6.98e6 22.97e6 12.10e6 
Mean: 10.88e6
Standard deviation: 5.36e6


In [134]:
lin_scores = cross_val_score(lin_reg, X_train_EU, y_train_EU,
                             scoring="r2", cv=10)
display_scores(lin_scores)

Scores: 0.747 0.243 0.805 0.497 0.888 0.685 0.953 0.556 0.667 0.362 
Mean: 0.64
Standard deviation: 0.227


On passe de RMSE=10.7e6 (baseline / N=-1) à 10.7e6 (N=10).  
On passe de R²=0.678 (baseline / N=-1) à :
- 0.650 (N=20),  
- 0.640 (N=15),  
- 0.640 (N=12),  
- 0.675 (N=10),  
- 0.672 (N=8),  
- 0.670 (N=5),  
- 0.676 (N=2),  
- 0.678 (N=1).
- 0.678 (N=0).

## Données recalibrées par StandardScaler

In [135]:
ss = StandardScaler()
X_train_EU_ss = ss.fit_transform(X_train_EU)
X_test_EU_ss = ss.transform(X_test_EU)

### Régression Ridge

In [136]:
ridge_ss = RidgeCV(alphas=np.logspace(-2, 2, 100), cv=10)    # 100 valeurs alpha entre 1e0 et 1e2
ridge_ss.fit(X_train_EU_ss, y_train_EU)

RidgeCV(alphas=array([1.00000000e-02, 1.09749877e-02, 1.20450354e-02, 1.32194115e-02, 1.45082878e-02, 1.59228279e-02, 1.74752840e-02, 1.91791026e-02, 2.10490414e-02,
       2.31012970e-02, 2.53536449e-02, 2.78255940e-02, 3.05385551e-02, 3.35160265e-02, 3.67837977e-02, 4.03701726e-02, 4.43062146e-02, 4.86260158e-02,
       5.33669923e-02, 5.85702082e-02, 6.42807312e-02, 7.05480231e-0...
       8.11130831e+00, 8.90215085e+00, 9.77009957e+00, 1.07226722e+01, 1.17681195e+01, 1.29154967e+01, 1.41747416e+01, 1.55567614e+01, 1.70735265e+01,
       1.87381742e+01, 2.05651231e+01, 2.25701972e+01, 2.47707636e+01, 2.71858824e+01, 2.98364724e+01, 3.27454916e+01, 3.59381366e+01, 3.94420606e+01,
       4.32876128e+01, 4.75081016e+01, 5.21400829e+01, 5.72236766e+01, 6.28029144e+01, 6.89261210e+01, 7.56463328e+01, 8.30217568e+01, 9.11162756e+01,
       1.00000000e+02]),
        cv=10)

In [137]:
best_alpha = ridge_ss.alpha_
best_alpha

0.01

In [138]:
# R²
R2_ss_ridge_cv = arrondi(ridge_ss.best_score_)
R2_ss_ridge_cv

0.64

On passe de 0.680 (baseline / N=-1) à :
- 0.650 (N=20).  
- 0.640 (N=15).  
- 0.676 (N=10).  
- 0.640 (N=12). 
- 0.673 (N=8).  
- 0.671 (N=5).
- 0.678 (N=2).  
- 0.680 (N=1).  
- 0.680 (N=0).  

##### R² sur le test set avec le meilleur alpha issu de la CV

In [139]:
ridge_ss = Ridge(alpha=best_alpha)
ridge_ss.fit(X_train_EU_ss, y_train_EU)
print(arrondi(ridge_ss.score(X_test_EU_ss, y_test_EU)))

0.801


On passe de 0.815 (baseline / N=-1) à :
- 0.773 (N=20).  
- 0.801 (N=15). 
- 0.801 (N=12).
- 0.84  (N=10).
- 0.838 (N=8).  
- 0.838 (N=5).  
- 0.836 (N=2).  
- 0.816 (N=1).  
- 0.815 (N=0).  

##### Conclusion sur la 3è approche :

L'utilisation des surfaces des PropertyUseType, en regroupant les types rares implique un effet défavorable sur le train set et favorable sur le test set.